File: crates/piston-nn/src/sinusoidal.rs
use piston::{rvec, Tensor};
use piston_macros::scoped_module;

use crate::Module;

/// Implements sinusoidal positional encodings as described in "Attention Is All You Need".
#[derive(Clone, Debug)]
pub struct SinusoidalEmbedding {
    #[allow(dead_code)]
    dim: usize,
    inv_freq: Tensor,
}

pub struct SinusoidalInput {
    pub input: Tensor,
    pub offset: usize,
}

impl SinusoidalEmbedding {
    pub fn new(dim: usize, device: &piston::Device) -> anyhow::Result<Self> {
        // Create position frequencies
        let mut freqs = Vec::with_capacity(dim / 2);
        for i in (0..dim).step_by(2) {
            freqs.push(1.0 / (10000f32.powf(i as f32 / dim as f32)));
        }

        // Create inverse frequency tensor
        let inv_freq = Tensor::from_data(&freqs, dim / 2, device.clone());

        Ok(Self { dim, inv_freq })
    }
}

#[scoped_module]
impl Module for SinusoidalEmbedding {
    type Input = SinusoidalInput;
    type Output = Tensor;

    fn schedule(&self, input: Self::Input) -> anyhow::Result<Self::Output> {
        let SinusoidalInput { input, offset } = input;

        // Get sequence length from input shape
        let seq_len = input.shape()[1];

        // Create position sequence [0, 1, 2, ..., seq_len-1]
        let pos_seq =
            Tensor::arange::<f32>(offset as f32, (offset + seq_len) as f32, input.device())?;

        // Compute outer product between positions and frequencies
        let sinusoid_inp =
            pos_seq
                .unsqueeze(1)?
                .matmul(self.inv_freq.clone().unsqueeze(0)?, false, false)?;

        // Compute sin and cos
        let sin = sinusoid_inp.clone().sin()?;
        let cos = sinusoid_inp.cos()?;

        let last_dim = sin.shape().len() - 1;
        // Interleave sin and cos values
        let pos_emb = Tensor::cat(rvec![sin, cos], last_dim)?;

        // Add batch dimension if needed
        let pos_emb = pos_emb.unsqueeze(0)?;

        // Add the positional embeddings to the input
        input + pos_emb
    }
}

File: packages/web/src/nn/sinusoidal.ts
import { arange, cat, tensor } from "@/globals";
import { Module } from "@/nn/module";
import { Buffer } from "@/nn/parameter";
import { Tensor } from "@/tensor";
import { Device } from "@/wasm";

/**
 * Implements sinusoidal positional encodings as described in "Attention Is All You Need".
 */
export class SinusoidalEmbedding extends Module {
  private dim: number;
  private invFreq: Buffer;

  /**
   * Creates a SinusoidalEmbedding module.
   * @param dim The embedding dimension. Must be even.
   * @param device Optional device to create the internal tensors on.
   */
  constructor(dim: number, device?: Device | "cpu") {
    super();
    if (dim % 2 !== 0) {
      throw new Error("Embedding dimension must be even.");
    }
    this.dim = dim;

    const freqs = new Float32Array(dim / 2);
    for (let i = 0; i < dim / 2; i++) {
      freqs[i] = 1.0 / Math.pow(10000, (2 * i) / dim);
    }

    this.invFreq = new Buffer(tensor(freqs, { device }));
  }

  /**
   * Applies sinusoidal positional embedding to the input tensor.
   * @param input The input tensor, expected shape [batch?, seqLen, dim].
   * @param offset The starting position offset for the sequence. Defaults to 0.
   * @returns Tensor with positional embeddings added.
   */
  forward(input: Tensor, offset: number = 0): Tensor {
    // Ensure invFreq is on the same device as input
    if (this.invFreq.device !== input.device) {
      throw new Error("invFreq must be on the same device as input");
    }

    const device = input.device;

    const seqLen = input.shape[input.shape.length - 2]; // Get sequence length (second to last dim)

    // Create position sequence [offset, offset + 1, ..., offset + seqLen - 1]
    const posSeq = arange({ start: offset, end: offset + seqLen, device });

    // Compute outer product: [seqLen, 1] @ [1, dim/2] -> [seqLen, dim/2]
    const sinusoidInp = posSeq.unsqueeze(1).matmul(this.invFreq.unsqueeze(0));

    // Compute sin and cos: [seqLen, dim/2]
    const sin = sinusoidInp.sin();
    const cos = sinusoidInp.cos();

    // Interleave sin and cos: [seqLen, dim]
    const posEmb = cat([sin, cos], { dim: -1 }); // Concatenate along the last dimension

    // Add batch dimension if input has batch: [1, seqLen, dim] or keep as [seqLen, dim]
    const posEmbFinal = input.shape.length === 3 ? posEmb.unsqueeze(0) : posEmb;

    // Add the positional embeddings to the input
    return input.add(posEmbFinal);
  }
}
